### ISSUE TITLE
colmem: batch capacity is capped by large rows

### ISSUE THREAD

**Issue by DrewKimball** [cockroachdb/cockroach#87587]
This issue tracks a known (and expected) problem with the current batch-size limiting behavior that can cause the batch capacity to continually decrease until it only contains one row. Currently, when a batch significantly exceeds its memory limit, we cap the maximum capacity at half the current capacity and reallocate. When this happens repeatedly due to large rows, the maximum capacity will shrink until it eventually reaches one row, which can significantly impact processing speed. Ideally we would have have a way to increase the maximum capacity of the batch when the rows are small in order to adapt to this situation.

An additive-increase/multiplicative-decrease scheme seems attractive at first for trying to use as much memory as is available while remaining within the memory limit, but is problematic because additive-increase would entail too many reallocations. However, consider the case when we have exceeded the memory limit and reallocate the batch - we could reallocate with the same capacity, but at first limit the amount we use to half of that. Then, if on the next iteration we do not exceed the limit (or get close to doing so), we can additively increase the amount of the batch's capacity that we allow to be used. @yuzefovich does this idea seem promising to you?

**Comment by yuzefovich** [comment]
I agree, this sounds like a good improvement to `colmem.SetAccountingHelper` (this would apply to `colmem.AccountingHelper` which works at the batch granularity since it doesn't have a way to "limit" the caller to use only a portion of the allocated capacity).

**Comment by DrewKimball** [comment]
I think it's already expected that the returned capacity can be different than the requested one even for `colmem.AccountingHelper`, right? All the callers immediately update their capacity with the capacity of the returned batch after calling `AccountingHelper.ResetMaybeReallocate`, so I think it could apply there too.

**Comment by DrewKimball** [comment]
Maybe we could do something like have a wrapper around `coldata.Batch` that just returns a smaller capacity than the wrapped batch has.

**Comment by yuzefovich** [comment]
I see, yeah, that should work, good point.

**Comment by mgartner** [comment]
@yuzefovich @michae2 @DrewKimball Do the benefits from this work provide high value for us in the near future, or should we backlog?

**Comment by yuzefovich** [comment]
I'd say this would provide a moderate value (it seems like we've only seen this in one support ticket), but it also doesn't seem like a large project.
